### ISSUE TITLE
`jupyter lab build` fails on NFS drives

### ISSUE THREAD

**Issue by neighthan** [jupyterlab/jupyterlab#4738]
I was following the instructions [here](https://github.com/jupyterlab/jupyterlab/blob/master/CONTRIBUTING.md) for setting up Jupyter lab for contributing, but when I run `jupyter lab build` I get

```
Traceback (most recent call last):
  File "/cluster/nhunt/anaconda/envs/jlab/bin/jupyter-lab", line 11, in <module>
    load_entry_point('jupyterlab', 'console_scripts', 'jupyter-lab')()
  File "/cluster/nhunt/anaconda/envs/jlab/lib/python3.6/site-packages/jupyter_core/application.py", line 266, in launch_instance
    return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
  File "/cluster/nhunt/anaconda/envs/jlab/lib/python3.6/site-packages/traitlets/config/application.py", line 658, in launch_instance
    app.start()
  File "/cluster/nhunt/anaconda/envs/jlab/lib/python3.6/site-packages/notebook/notebookapp.py", line 1605, in start
    super(NotebookApp, self).start()
  File "/cluster/nhunt/anaconda/envs/jlab/lib/python3.6/site-packages/jupyter_core/application.py", line 255, in start
    self.subapp.start()
  File "/cluster/nhunt/github/jupyterlab/jupyterlab/labapp.py", line 64, in start
    command=command, logger=self.log)
  File "/cluster/nhunt/github/jupyterlab/jupyterlab/commands.py", line 278, in build
    command=command, clean_staging=clean_staging)
  File "/cluster/nhunt/github/jupyterlab/jupyterlab/commands.py", line 428, in build
    clean=clean_staging
  File "/cluster/nhunt/github/jupyterlab/jupyterlab/commands.py", line 865, in _populate_staging
    shutil.copytree(pjoin(HERE, 'staging', 'templates'), templates)
  File "/cluster/nhunt/anaconda/envs/jlab/lib/python3.6/shutil.py", line 359, in copytree
    raise Error(errors)
shutil.Error: [('/cluster/nhunt/github/jupyterlab/jupyterlab/staging/templates', '/cluster/nhunt/anaconda/envs/jlab/share/jupyter/lab/staging/templates%', "[Errno 22] Invalid argument: '/cluster/nhunt/anaconda/envs/jlab/share/jupyter/lab/staging/templates'")]```
```

This is in a fresh `conda` environment that I created just before starting the steps to set up Jupyter lab on Ubuntu 16.04. Any idea why this is failing/how to fix it?

EDIT: here's the full script that I'm running

```bash
conda create -n jlab -y
source activate jlab
conda install -c conda-forge nodejs -y
conda install -c conda-forge notebook -y
git clone https://github.com/jupyterlab/jupyterlab.git
cd jupyterlab
pip install -e .
jlpm install
jlpm run build  # Build the dev mode assets (optional)
jlpm run build:core  # Build the core mode assets (optional)
jupyter lab build  # Build the app dir assets (optional)
```

I'm not sure what the issue was with trying to copy the templates directory - it looks like an existing directory should be removed and then the templates copied over, but I have both the template directories mentioned in the error message, despite `copytree` failing, so it seems there's a problem with `rmtree`. For now, I have this working by just wrapping the `copytree` with a `try`/`except` instead of removing an existing templates directory.

**Comment by behinger** [comment]
I have the same problem with pip-installed jupyterlab (jupyterlab==0.33.11)
During installation of a jupyter widget

```
jupyter labextension install jupyter-matplotlib --debug
```
The same copytree error appeared.
```
Traceback (most recent call last):

  File "/net/store/nbp/users/behinger/projects/etcomp/local/etcompvenv/lib/python3.5/site-packages/jupyterlab/labextensions.py", line 83, in start
    logger=self.log, command=command)

  File "/net/store/nbp/users/behinger/projects/etcomp/local/etcompvenv/lib/python3.5/site-packages/jupyterlab/commands.py", line 278, in build
    command=command, clean_staging=clean_staging)

  File "/net/store/nbp/users/behinger/projects/etcomp/local/etcompvenv/lib/python3.5/site-packages/jupyterlab/commands.py", line 449, in build
    clean=clean_staging

  File "/net/store/nbp/users/behinger/projects/etcomp/local/etcompvenv/lib/python3.5/site-packages/jupyterlab/commands.py", line 886, in _populate_staging
    shutil.copytree(pjoin(HERE, 'staging', 'templates'), templates)

  File "/net/store/nbp/users/behinger/projects/etcomp/local/etcompvenv/lib/python3.5/shutil.py", line 353, in copytree
    raise Error(errors)

shutil.Error: [('/net/store/nbp/users/behinger/projects/etcomp/local/etcompvenv/lib/python3.5/site-packages/jupyterlab/staging/templates', '/net/store/nbp/users/behinger/projects/etcomp/local/etcompvenv/share/jupyter/lab/staging/templates', "[Errno 22] Invalid argument: '/net/store/nbp/users/behinger/projects/etcomp/local/etcompvenv/share/jupyter/lab/staging/templates'")]


````
The hotfix of @neighthan worked for me.

**Comment by blink1073** [comment]
We don't want to keep possibly stale `templates` in there, so I do think we want to remove the old directory and refresh it.  The `[Errno 22]` error seems to be related to permissions from some googling on the subject.

**Comment by neighthan** [comment]
I had originally thought that something was wrong with the `shutil.rmtree` call as well because I have both `templates` directories when I get this error. But upon further investigation, it seems that `shutil.rmtree` is working fine and so is `shutil.copytree` (ish). That is, the new templates directory is indeed copied over but the above error is still thrown. So, I've changed my workaround to the following (around line 878 of `jupyterlab/jupyterlab/commands.py`):

```python
try:
    shutil.copytree(pjoin(HERE, 'staging', 'templates'), templates)
except shutil.Error as error:
    if '[Errno 22] Invalid argument' not in error.args[0][0][2]:
        raise
```

I also found this [Python bug report](https://bugs.python.org/issue24564) which I think is the issue we're having. It only applies when copying to + from an NFS-formatted drive (which mine is). @behinger, is that the case for you as well? Here are some relevant snippets:

> On a bit of further investigation, the NFS files have an xattr `system.nfs4_acl`. This can be read, but attempting to write it fails with EINVAL. Attempting to copy from NFS to non-NFS fails with ENOTSUP, which is caught and ignored, but copying from NFS to NFS raises EINVAL, which raises. Adding `EINVAL` to the ignored errnos would fix the problem, but might hide real failures (I'm not sure about the real failures, but it seems logical).

> I think this is an acceptable compromise considering that:
> 1. because of this the copy operation will ultimately fail (despite file will be copied)
> 2. there is nothing we can do except emit a warning somehow 
> 3. EPERM and ENODATA are already silently ignored (also ENOTSUP but that is more legit)
> 
> For completeness, other than copytree() also copy2() and move() are affected. 
  I submitted a PR: https://github.com/python/cpython/pull/8601.

So, it seems like this should be fixed at some point but the consensus may be that it's okay to ignore? If we were to make a change in jupyterlab to fix this in the meantime, though, I'm not sure if my method above is robust enough (i.e. does Errno 22 only pertain to this specific error? and accessing that part of the error string in a way that won't throw errors if we've caught a different error than here).

**Comment by behinger** [comment]
Yes, it is a nfs drive - nice find! I do not have the ressources right now to make tests on my own. But I'm happy to run some code or answer questions to our setup if needed.

Thanks for the digging!

**Comment by blink1073** [comment]
@neighthan, does that work with the simplified:

```python
try:
    shutil.copytree(pjoin(HERE, 'staging', 'templates'), templates)
except shutil.Error as error:
    if error.errno != 2:   # cf https://bugs.python.org/issue24564
        raise
```

**Comment by neighthan** [comment]
@blink1073 I did want to use `error.errno` originally, but unfortunately I find that it's `None` when I get this error. One improvement, though, so we don't get more errors if `error.args[0][0][2]` doesn't exist would be to check `if "[Errno 22] Invalid argument" in str(error)` (or just "[Errno 22]"?). I suppose another route would be that if we catch an error, we check whether the `templates` directory now exists and if it has files that match those in the new templates directory that we just tried to copy over. If the directory exists and the files match, then we don't raise the caught error again; otherwise, we do.
